File: mbfpp/code/ScalaExamples/src/main/scala/com/mblinn/mbfpp/oo/iterator/HigherOrderFunctions.scala
package com.mblinn.mbfpp.oo.iterator

object HigherOrderFunctions {

  def sumSequence(sequence : Seq[Int]) =
    if(sequence.isEmpty) 0 else sequence.reduce((acc, curr) => acc + curr)
  
  def prependHello(names : Seq[String]) =
    names.map((name) => "Hello, " + name)
    
  val isVowel = Set('a', 'e', 'i', 'o', 'u')
  def vowelsInWord(word: String) = word.filter(isVowel).toSet
    
}

File: mbfpp/code/ClojureExamples/src/mbfpp/oo/iterator/higher_order_functions.clj
(ns mbfpp.oo.iterator.higher-order-functions)

(defn sum-sequence [s]
  {:pre [(not (empty? s))]}
  (reduce + s))

(defn prepend-hello [names]
  (map (fn [name] (str "Hello, " name)) names))

(def vowel? #{\a \e \i \o \u})
(defn vowels-in-word [word]
  (set (filter vowel? word)))
